crates/sui-framework/sources/collectible/kiosk.move

```diff
@@ -0,0 +1,474 @@
+/// that has issued `TransferPolicyCap` in a Kiosk. To do so, they need
+/// to make an offer, and any party can purchase the item for the amount of
+/// SUI set in the offer. The responsibility to follow the transfer policy
+/// set by the creator of the `T` is on the buyer.
+///
+/// 3. Marketplaces: marketplaces can either watch for the offers made in
+/// personal Kiosks or even integrate the Kiosk primitive and build on top
+/// of it. In the custom logic scenario, the `TransferPolicyCap` can also
+/// be used to implement application-specific transfer rules.
+///
+module sui::kiosk {
+    use std::option::{Self, Option};
+    use sui::object::{Self, UID, ID};
+    use sui::dynamic_object_field as dof;
+    use sui::dynamic_field as df;
+    use sui::package::{Self, Publisher};
+    use sui::tx_context::{TxContext, sender};
+    use sui::balance::{Self, Balance};
+    use sui::coin::{Self, Coin};
+    use sui::sui::SUI;
+    use sui::event;
+
+    // Collectible is a special case to avoid storing `Publisher`.
+    friend sui::collectible;
+
+    /// For when trying to withdraw profits as owner and owner is not set.
+    const EOwnerNotSet: u64 = 0;
+    /// For when trying to withdraw profits and sender is not owner.
+    const ENotOwner: u64 = 1;
+    /// For when Coin paid does not match the offer price.
+    const EIncorrectAmount: u64 = 2;
+    /// For when incorrect arguments passed into `switch_mode` function.
+    const EIncorrectArgument: u64 = 3;
+    /// For when Transfer is accepted by a wrong Kiosk.
+    const EWrongTarget: u64 = 4;
+    /// For when trying to withdraw higher amount than stored.
+    const ENotEnough: u64 = 5;
+    /// For when trying to close a Kiosk and it has items in it.
+    const ENotEmpty: u64 = 6;
+
+    /// An object that stores collectibles of all sorts.
+    /// For sale, for collecting reasons, for fun.
+    struct Kiosk has key, store {
+        id: UID,
+        /// Balance of the Kiosk - all profits from sales go here.
+        profits: Balance<SUI>,
+        /// Always point to `sender` of the transaction.
+        /// Can be changed by calling `set_owner` with Cap.
+        owner: address,
+        /// Number of items stored in a Kiosk. Used to allow unpacking
+        /// an empty Kiosk if it was wrapped or has a single owner.
+        item_count: u32
+    }
+
+    /// A capability that is issued for Kiosks that don't have owner
+    /// specified.
+    struct KioskOwnerCap has key, store {
+        id: UID,
+        for: ID
+    }
+
+    /// A "Hot Potato" forcing the buyer to get a transfer permission
+    /// from the item type (`T`) owner on purchase attempt.
+    struct TransferRequest<phantom T: key + store> {
+        /// Amount of SUI paid for the item. Can be used to
+        /// calculate the fee / transfer policy enforcement.
+        paid: u64,
+        /// The ID of the Kiosk the object is being sold from.
+        /// Can be used by the TransferPolicy implementors to ban
```

Comment:
The banning use-case does not make a lot of sense to me, since the owner of a banned kiosk can always just create a new one whose ID is not in the denylist. Explaining that this is useful as an allowlist for special-casing some kiosks seems better (and I love that idea--powerful stuff!)